Firestore/Example/Tests/Integration/FSTTransactionTests.mm

```diff
@@ -20,184 +20,346 @@
+      [expectation fulfill];
+    }];
+
+    [_testCase awaitExpectations];
+  }
+}
+
+- (void)runSuccessfulTransaction {
+  XCTestExpectation *expectation = [_testCase expectationWithDescription:@"runTransaction"];
+  [_db
       runTransactionWithBlock:^id _Nullable(FIRTransaction *transaction, NSError **error) {
-        [transaction getDocument:doc error:error];
-        XCTAssertNil(*error);
+        for (TransactionStage stage in self->_stages) {
+          stage(transaction, self->_docRef);
+        }
         return @YES;
       }
       completion:^(id _Nullable result, NSError *_Nullable error) {
-        XCTAssertNil(result);
-        // We currently require every document read to also be written.
-        // TODO(b/34879758): Fix this check once we drop that requirement.
-        XCTAssertNotNil(error);
-        XCTAssertEqual(error.code, FIRFirestoreErrorCodeInvalidArgument);
         [expectation fulfill];
+        NSString *message =
+            [NSString stringWithFormat:@"Expected the sequence %@, to succeed, but got %ld.",
+                                       [self listStages], [error code]];
+        [self->_testCase assertNilErrorWithMessage:message error:error];
       }];
-  [self awaitExpectations];
-}
 
-- (void)testDeleteDocument {
-  FIRFirestore *firestore = [self firestore];
-  FIRDocumentReference *doc = [[firestore collectionWithPath:@"towns"] documentWithAutoID];
-  [self writeDocumentRef:doc data:@{@"foo" : @"bar"}];
-  FIRDocumentSnapshot *snapshot = [self readDocumentForRef:doc];
-  XCTAssertEqualObjects(@"bar", snapshot[@"foo"]);
+  [_testCase awaitExpectations];
+}
 
-  XCTestExpectation *expectation = [self expectationWithDescription:@"transaction"];
-  [firestore
+- (void)runFailingTransaction:(FIRFirestoreErrorCode)expected {
+  XCTestExpectation *expectation = [_testCase expectationWithDescription:@"runTransaction"];
+  [_db
       runTransactionWithBlock:^id _Nullable(FIRTransaction *transaction, NSError **error) {
-        [transaction deleteDocument:doc];
+        for (TransactionStage stage in self->_stages) {
+          stage(transaction, self->_docRef);
+        }
         return @YES;
       }
       completion:^(id _Nullable result, NSError *_Nullable error) {
-        XCTAssertEqualObjects(@YES, result);
-        XCTAssertNil(error);
         [expectation fulfill];
+        NSString *message =
+            [NSString stringWithFormat:@"Expected the sequence (%@), to fail, but it didn't.",
+                                       [self listStages]];
+        [self->_testCase assertErrorWithMessage:message error:error];
       }];
-  [self awaitExpectations];
 
-  snapshot = [self readDocumentForRef:doc];
+  [_testCase awaitExpectations];
+}
+
+- (void)cleanupTester {
+  _stages = [NSArray array];
+  // Set the docRef to something else to lose the original reference.
+  _docRef = [[self->_db collectionWithPath:@"reset"] documentWithAutoID];
+}
+
+- (NSString *)listStages {
```

Comment:
See the naming guide again. Methods of this form should be a noun describing what's returned. `stageNames`?